### PR TITLE
Per 7292 add cross origin iframe support

### PR DESCRIPTION
**Pull request by aryanku-dev**

### CODE REVIEW

**Review by rishigupta1599 — COMMENTED** [review]
Nice adaptation of the cross-origin iframe pattern from percy-playwright, and the closed shadow root support via CDP is a genuinely useful addition. Posting detailed inline comments separately.

File: index.js [hunk 1]
```diff
@@ -5,24 +5,241 @@ const sdkPkg = require('./package.json');
...
+    /* eslint-disable-next-line no-undef */
+    return PercyDOM.serialize(opts);
+  }, { ...options, enableJavascript: true });
+  log.debug(`Serialized cross-origin iframe: ${frameUrl}`);
+
+  // Get the iframe's element data from the main page context
+  /* istanbul ignore next: browser-executed evaluation function */
+  const iframeData = await page.evaluate((fUrl) => {
+    const iframes = Array.from(document.querySelectorAll('iframe'));
+    const matchingIframe = iframes.find(iframe => iframe.src.startsWith(fUrl));
```
Comment (rishigupta1599) [inline index.js]: Matching by `iframe.src.startsWith(frame.url())` is brittle in realistic pages:

- Multiple iframes pointing at the same origin (e.g. two ads from the same provider) will all match the first one; the wrong `percyElementId` gets attributed.
- An iframe that navigated client-side will have `iframe.src` (the original attribute) differ from `frame.url()` (the live URL). No match, the frame silently drops.
- Iframes without a `src` attribute (srcdoc, programmatically set via `contentWindow.location`) always return `''` for `iframe.src`.
- The `startsWith` direction is also suspect: if `iframe.src = 'https://other.com/'` but `frame.url() = 'https://other.com/embed'` after a redirect, `'https://other.com/'.startsWith('https://other.com/embed')` is false.

Consider matching via the frame's owning element — puppeteer exposes `frame.frameElement()` in recent versions, or use `CDP DOM.getFrameOwner({frameId})` to get the exact backendNodeId and map back to `data-percy-element-id` deterministically. Inherited from percy-playwright, but worth fixing.

File: index.js [hunk 2]
```diff
@@ -5,24 +5,241 @@ const sdkPkg = require('./package.json');
...
+// Processes a single cross-origin frame to capture its snapshot and resources.
+async function processFrame(page, frame, options, percyDOM) {
+  const frameUrl = frame.url();
+  log.debug(`Processing cross-origin iframe: ${frameUrl}`);
+
+  /* istanbul ignore next: browser-executed iframe serialization */
+  const iframeSnapshot = await frame.evaluate((opts) => {
+    /* eslint-disable-next-line no-undef */
+    return PercyDOM.serialize(opts);
+  }, { ...options, enableJavascript: true });
```
Comment (rishigupta1599) [inline index.js]: The spread order silently overrides a user-provided `enableJavascript: false` for iframe serialization. For top-level DOM serialization on line 66-69 the user's option is respected; for cross-origin iframes it isn't. If this is intentional (percy-playwright has a comment explaining it), please carry that comment over here — otherwise users will be confused why their snapshot options appear to have an effect on the main page but not on cross-origin iframes.

File: index.js [hunk 3]
```diff
@@ -5,24 +5,241 @@ const sdkPkg = require('./package.json');
...
+    injectableFrames.map(frame =>
+      processFrame(page, frame, options, percyDOM).catch(e => {
+        log.debug(`Failed to process cross-origin frame ${frame.url()}: ${e.message}`);
+        return null;
+      })
+    )
+  )).filter(Boolean);
+
+  if (processedFrames.length > 0) {
+    domSnapshot.corsIframes = processedFrames;
```
Comment (rishigupta1599) [inline index.js:247]: percy-playwright unconditionally assigns `domSnapshot.corsIframes = processedFrames;` (even when empty). Here it's only set when non-empty. If the CLI/DOM processor branches on `Array.isArray(domSnapshot.corsIframes)` vs `'corsIframes' in domSnapshot`, this behavioral divergence can surface bugs that don't repro on playwright. Worth either (a) matching playwright's behavior to keep the SDK surface uniform, or (b) confirming the CLI tolerates either shape and noting the intent in a comment.

File: index.js [hunk 4]
```diff
@@ -5,24 +5,241 @@ const sdkPkg = require('./package.json');
...
+    log.debug(`Captured ${processedFrames.length} cross-origin iframe(s)`);
+  }
+
+  // Capture cookies
+  const cookies = await page.cookies();
```
Comment (rishigupta1599) [inline index.js]: `page.cookies()` was deprecated in Puppeteer v23 in favor of `page.browserContext().cookies()` / `browser.cookies()` and is slated for removal. If the SDK supports puppeteer >= 23 (check the peerDependency in package.json), this will emit deprecation warnings for every snapshot. Worth doing a version-gated call or at least documenting the minimum puppeteer version now required.

Also note: this captures every cookie on the page, including HttpOnly/Secure session cookies, and ships them over to the CLI. Same behavior as percy-playwright, but double-check that the CLI scrubs/only-uses these for fetching iframe resources and doesn't persist them in any artifact.

File: index.js [hunk 5]
```diff
@@ -5,24 +5,241 @@ const sdkPkg = require('./package.json');
...
+];
+
+function isUnsupportedIframeSrc(src) {
+  /* istanbul ignore next: defensive guard — callers already filter falsy URLs */
+  if (!src) return true;
+  return UNSUPPORTED_IFRAME_SRCS.some(prefix => src === prefix || src.startsWith(prefix));
+}
+
+// Processes a single cross-origin frame to capture its snapshot and resources.
+async function processFrame(page, frame, options, percyDOM) {
```
Comment (rishigupta1599) [inline index.js]: `percyDOM` is accepted as a parameter here but never referenced inside the function — injection already happens in `captureSerializedDOM` before `processFrame` is called. Either drop the unused parameter or actually use it (e.g., as a fallback re-inject if the first injection got wiped by an iframe navigation between inject + serialize).

File: index.js [hunk 6]
```diff
@@ -5,24 +5,241 @@ const sdkPkg = require('./package.json');
...
+      const { object: hostObj } = await client.send('DOM.resolveNode', {
+        backendNodeId: pair.hostBackendNodeId
+      });
+      const { object: shadowObj } = await client.send('DOM.resolveNode', {
+        backendNodeId: pair.shadowBackendNodeId
+      });
+
+      /* istanbul ignore next: CDP-injected function */
+      await client.send('Runtime.callFunctionOn', {
+        functionDeclaration: 'function(shadowRoot) { window.__percyClosedShadowRoots.set(this, shadowRoot); }',
```
Comment (rishigupta1599) [inline index.js:372]: This is a real bug for pages where the closed shadow root lives inside a (cross- or same-origin) iframe.

`DOM.getDocument({depth: -1, pierce: true})` pierces iframe boundaries, so `closedPairs` can include host nodes that belong to a child frame's document. `DOM.resolveNode` resolves those nodes into *that frame's* JS execution context, which means `Runtime.callFunctionOn` with the resulting `objectId` runs in the child frame's context — not the top-level page where `window.__percyClosedShadowRoots` was created on line 192-194.

Consequences:
1. On the child-frame run, `window.__percyClosedShadowRoots` is `undefined`, the `.set(...)` throws `TypeError`, and the outer `catch` on line 215 aborts the remainder of the loop — subsequent closed shadow roots (even ones on the top-level document) never get exposed.
2. Even if you guarded against the throw, the WeakMap would live on the wrong window and `PercyDOM.serialize` on the main page wouldn't find it.

Options:
- Pass `executionContextId` explicitly to `Runtime.callFunctionOn` targeting the top frame's executionContext (track via `Runtime.enable` + `executionContextCreated` events).
- Skip closed pairs whose host isn't in the main frame (track which nodes came from which document via the `getDocument` response's `frameId`/`documentURL`) — safer first step.
- Or, if you want to also capture closed shadows inside iframes, create the WeakMap in each relevant execution context.

File: index.js [hunk 7]
```diff
@@ -5,24 +5,241 @@ const sdkPkg = require('./package.json');
...
+  } catch {
+    // Not a Chromium browser — skip silently
+    return;
+  }
+
+  try {
+    await client.send('DOM.enable');
+
+    // Get the full DOM tree, piercing all shadow roots including closed ones
+    const { root } = await client.send('DOM.getDocument', {
```
Comment (rishigupta1599) [inline index.js:311]: `DOM.getDocument({depth: -1, pierce: true})` on every single snapshot is a meaningful cost for large apps — the full DOM + every iframe + every shadow tree gets serialized and shipped over CDP, even when there are zero closed shadow roots on the page. For a heavy SPA this can add hundreds of ms and a non-trivial CPU spike per snapshot.

Consider a cheap pre-check first: e.g., monkey-patch `Element.prototype.attachShadow` inside PercyDOM injection so any call with `{mode:'closed'}` sets a `window.__percyHasClosedShadow = true` flag, then only pay the `pierce: true` cost when that flag is set. That keeps the common case (no closed shadows) free.

File: index.js (hunk 1 above)
Comment (aryanku-dev) [reply index.js]: Acknowledged. Added a comment documenting the `iframe.src.startsWith(frame.url())` limitation and noting `frame.frameElement()` / `CDP DOM.getFrameOwner` as better alternatives for a future improvement.

File: index.js (hunk 2 above)
Comment (aryanku-dev) [reply index.js]: Fixed. Added a comment explaining why `enableJavascript: true` is intentionally forced — it prevents the standard iframe serialization logic from running since we handle cross-origin iframe serialization manually.

File: index.js (hunk 3 above)
Comment (aryanku-dev) [reply index.js:247]: Fixed. Now unconditionally assigning `domSnapshot.corsIframes = processedFrames` to match percy-playwright's behavior and keep the SDK surface uniform.

File: index.js (hunk 4 above)
Comment (aryanku-dev) [reply index.js]: Added a comment noting `page.cookies()` is deprecated in Puppeteer v23+ and that cookies are only used by the CLI for fetching cross-origin iframe resources during asset discovery.

File: index.js (hunk 5 above)
Comment (aryanku-dev) [reply index.js]: Fixed. Removed the unused `percyDOM` parameter from `processFrame` and updated all call sites.

File: index.js (hunk 6 above)
Comment (aryanku-dev) [reply index.js:372]: Fixed. Added `if (node.contentDocument) return;` at the top of `walkNodes()` to skip iframe boundary nodes. This prevents traversing into child frame documents where `window.__percyClosedShadowRoots` doesn't exist. Added a test case that verifies closed shadow roots inside iframe contentDocument nodes are skipped.

File: index.js (hunk 7 above)
Comment (aryanku-dev) [reply index.js:311]: Added a performance note comment documenting this as a known consideration. The monkey-patch `attachShadow` pre-check optimization is a good idea for a follow-up — it requires changes to PercyDOM injection which is shared across SDKs.

**Comment by aryanku-dev** [comment]
## 🤖 stack:pr-review — Automated review (BrowserStack AI harness)

**Summary:** Adds cross-origin (and nested) iframe capture via a flat `page.frames()` walk + per-frame `PercyDOM.serialize`, plus closed-shadow-DOM capture through a CDP `pierce:true` traversal exposing closed roots to a page-side WeakMap. Includes depth-cap, cycle guard, `data-percy-ignore`/`ignoreIframeSelectors`, and Puppeteer-v23+ cookie fallback.

| Priority | Category | Check | Status |
|---|---|---|---|
| High | Security | No hardcoded secrets | Pass |
| High | Security | Input validation | Pass — URLs guarded by try/catch + `new URL`; selectors run in try/catch |
| High | Security | AuthZ / IDOR / SQLi | N/A |
| High | Correctness | Logic & edge cases | Pass — depth cap, A→B→A cycle guard, cross-origin-vs-immediate-parent check, exact-then-normalized src match; mirrors canonical `percy-playwright` |
| High | Correctness | Explicit error handling | Pass — per-frame `.catch` (debug), top-level catch (error); cookie/CDP failures non-fatal by design |
| High | Correctness | No race conditions | Pass — `Promise.all` fan-outs independent; CDP serialized on one client |
| Medium | Testing | New code / error paths tested | Pass — 53 specs incl. non-Chromium, CDP throw, child-frame guard, nested shadow, invalid URL |
| Medium | Testing | No regressions | Pass — CI green; local failures pre-existing env issue (see note) |
| Medium | Performance | No unbounded fetch | Pass — frame walk bounded by hard cap 25; `pierce:true` cost documented |
| Medium | Quality | Patterns / focused | Pass — helpers isolated in `_iframe_shim.js` |
| Low | Quality | Names / comments / deps | Pass |

**Findings:** no code Fail. Verified the prior-pass concerns: the `contentDocument` guard in `walkNodes` ensures collected closed-shadow hosts always live in the top frame, so `DOM.resolveNode` resolves them into the same main world as the WeakMap (asserted by the "skips closed shadow roots inside child frame documents" spec); the `pierce:true` note is documentation-only over a correct single scoped traversal.

**Verification:** Lint Pass; Typecheck (`tsd`) Pass; all PR CI checks SUCCESS. Local specs fail only with `ECONNREFUSED ::1:5338` (local `node_modules` lacks `@percy/cli`, so the mock server never starts — `master` fails identically). Tests are genuine/unweakened (no `xit`/`fit`/skips). Changes at `1ea4d14` already on the branch.

### Overall: ✅ **Pass** — Approve. No human-decision blockers. (Consider re-requesting review so the `blocked` mergeable_state clears.)

<sub>Posted by the BrowserStack `stack:pr-review` harness.</sub>

**Comment by aryanku-dev** [comment]
# Claude Code PR Review

**PR:** https://github.com/percy/percy-puppeteer/pull/960  •  **Head:** bb0c123  •  **Reviewers:** stack:code-reviewer

## Summary
Adds cross-origin iframe capture (`corsIframes`) and closed-shadow-DOM exposure via CDP to `percySnapshot`, consuming canonical iframe helpers from `@percy/sdk-utils@1.32.3-beta.1`; thoroughly tested, no blocking issues.

## Review Table

| Priority | Category | Check | Status | Notes |
|----------|----------|-------|--------|-------|
| High | Security | No hardcoded secrets | Pass | No credentials/tokens in diff. |
| High | Security | Authentication/authorization | N/A | No auth surface; SDK snapshot path only. |
| High | Security | Input validation | Pass | URLs parsed via `new URL` in try/catch; invalid selectors guarded; `isUnsupportedIframeSrc` scheme filter. |
| High | Security | No IDOR | N/A | No object-id access control in scope. |
| High | Security | No SQL injection | N/A | No SQL/DB access. |
| High | Correctness | Logic/edge cases | Pass | Depth cap, cycle detection, missing-id, invalid-URL, opaque-origin, v23+ cookie fallback all handled and tested. |
| High | Correctness | Explicit error handling | Pass | Per-frame inject/process wrapped in `.catch`; CDP and cookie collection best-effort with try/catch; snapshot never aborts. |
| High | Correctness | No race conditions | Pass | `Promise.all` fan-outs are independent; results keyed per-frame via Map; no shared mutable state. |
| Medium | Testing | New code has tests | Pass | 40+ specs across CORS iframe and closed-shadow handling; helpers unit-tested directly. |
| Medium | Testing | Error/edge cases tested | Pass | Injection failure, serialize failure, cookie-API throw, non-Chromium CDP, nested/cyclic/deep frames all covered. |
| Medium | Testing | Existing tests pass | Pass | Existing `percySnapshot` suite untouched; payload shape unchanged when no CORS frames. |
| Medium | Performance | No N+1/unbounded loops | Pass | Cheap filters applied before expensive parent.evaluate round-trips; `maxFrameDepth` cap bounds work. |
| Medium | Performance | Long tasks backgrounded | N/A | Synchronous-style snapshot; no long-running job to background. |
| Medium | Quality | Follows existing patterns | Pass | Mirrors percy-nightwatch/percy-playwright `captureSerializedDOM`; consumes shared sdk-utils helpers. |
| Medium | Quality | Focused change | Pass | Scoped to CORS-iframe + closed-shadow capture and the expected sdk-utils bump. |
| Low | Quality | Meaningful names / no dead code | Pass | Clear names; no commented-out/dead code; istanbul-ignore annotations justified. |
| Low | Quality | Comments explain why | Pass | Comments explain rationale (parent-DOM lookup, pierce scoping, startsWith pitfall, payload-shape stability). |
| Low | Quality | No unnecessary deps | Pass | New transitive deps arrive via the expected sdk-utils beta bump, all from the official yarn registry; no direct deps added beyond the version pin. |

## Findings
No blocking issues found.

Non-blocking observations (informational only):
- **File:** `index.js` (`processFrame`) — **Severity:** Low — **Reviewer:** stack:code-reviewer — **Issue:** The per-frame `PercyDOM.serialize` runs before the parent-frame `percyElementId` lookup, so a frame later dropped for a missing `data-percy-element-id` has already paid the serialize cost. **Suggestion:** Optionally resolve `percyElementId` before serializing to skip wasted work; minor, capped by `maxFrameDepth`.
- **File:** `package.json` — **Severity:** Low — **Reviewer:** stack:code-reviewer — **Issue:** `@percy/sdk-utils` pinned to exact `1.32.3-beta.1` (no caret). **Suggestion:** Expected for a beta per PR context; confirm it is relaxed to a range at GA.

---

**Verdict: PASS**

**Review by pranavz28 — APPROVED** [review]
